Refactor confirm approve page

Gudahtt opened the pull request:
The `confirm-approve` page has been converted from a set of container/component components to a functional component. The `withTokenTracker` higher-order component has been deleted, as it was replaced by the `useTokenTracker` hook.

Gudahtt committed:
Refactor confirm approve page

The `confirm-approve` page has been converted from a set of container/
component components to a functional component. The `withTokenTracker`
higher-order component has been deleted, as it was replaced by the
`useTokenTracker` hook.

Gudahtt commented on ui/app/pages/confirm-approve/confirm-approve.js:
```diff
@@ -0,0 +1,138 @@
+        setCustomPermissionAmount(tokenAmount)
+      }
+    },
+    [tokenAmount]
```
This would fail the 'exhaustive deps' rule, but I do want this to only run when `tokenAmount` changes. I could workaround that by using a ref, but that seems like a pain, for no benefit 🤔 

Not sure what convention we want to use here. i.e. maybe we should add a comment explaining why we're not exhaustively checking all deps whenever any are missing.

whymarrh replied:
A comment works, sure

brad-decker replied:
https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects see the yellow note field. If you do this your values will be stale except when `tokenAmount` changes. The "state" will be a stale object in memory that points to the value it was when the effect was first created. It'll be attached to a react fiber that also holds references to its previous render, and it'll be referenced by a future render. React has done work to mitigate this memory leak but its not rock solid.

I highly recommend following exhaustive deps always.

brad-decker replied:
https://github.com/facebook/react/issues/16087

Gudahtt replied:
It seems perfectly fine for the values to be stale if `tokenAmount` doesn't change - that's OK, right? Unless I'm missing something.

The memory leak is concerning though 👀

Gudahtt replied:
Reading up on their "better solutions to avoid re-running effects too often", none of them seem appreciably better in this situation. Except maybe that it avoids this memory leak, which I don't quite understand yet.

Gudahtt replied:
After talking with @brad-decker about this more and hearing him explain the last time he encountered the linked memory leak issue, I now agree that it would be better to err on the side of caution and just do what the `exhaustive-deps` rule says.

Gudahtt committed:
Add comment explaining why a dependency was omitted

Gudahtt committed:
Use ref instead of omitting dependency

Instead of omitting a dependency from the `useEffect` dependency array
to prevent it from running when the custom permission amount changed,
`useRef` is now used to explicitly track whether `tokenAmount` has
changed. This is a bit more verbose than the last version, but it
follows the `exhaustive-deps` rule/convention and it avoids some
potential memory leak problems.

brad-decker reviewed (approved):
LGTM
